### PR TITLE
Improve performance of resolving manifest file paths

### PR DESCRIPTION
**Pull request by anlaital-oura**
### Short description 📝

Running `tuist edit` in a large git repository is quite slow at the moment. Running Time Profiler indicates that globing is the root cause of the slowness as it globs the entire repository. This is despite `.tuistignore` being present, as exclusion is done only after globing.

These changes reduce the time it takes to run `tuist edit` in our large repository from 31s to ~3s~ ~1.5s~ 0.27s in release mode (M2 Max). Related discussion on [Tuist Slack](https://tuistapp.slack.com/archives/CCYNEGY1L/p1706843771208039).

I tried running `mise run tuist:test` but it fails acceptance tests even on a clean checkout. Could you advise on how tests should be run locally? I don't know if it's because of Tuist 4 changes, but a lot of the documentation seems a bit outdated on this.

**Before:**

image

**After:**

image

### How to test the changes locally 🧐

Run `tuist` scheme with `edit` argument against a custom working directory that is a large git repository.

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint:fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [x] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistLoader/Utils/ManifestFilesLocator.swift [hunk 1]
```diff
@@ -101,9 +101,21 @@ public final class ManifestFilesLocator: ManifestFilesLocating {
...
-        let tuistManifestsFilePaths = fileHandler.glob(path, glob: "**/*.swift")
-            .filter { fileNamesCandidates.contains($0.basename) }
-            .filter { hasValidManifestContent($0) }
+
+        var tuistManifestsFilePaths = [AbsolutePath]()
+
+        let enumerator = FileManager.default.enumerator(atPath: path.pathString)
+        while let candidatePath = enumerator?.nextObject() as? String {
+            let candidateURL = URL(fileURLWithPath: candidatePath)
+            if fileNamesCandidates.contains(candidateURL.lastPathComponent) {
```
Comment (danyf90) [inline Sources/TuistLoader/Utils/ManifestFilesLocator.swift:110]: :nit: not sure how relevant it is for performance, but I guess filtering out all non `.swift` files can be done without converting to URL

File: Sources/TuistLoader/Utils/ManifestFilesLocator.swift [hunk 2]
```diff
@@ -101,9 +101,21 @@ public final class ManifestFilesLocator: ManifestFilesLocating {
...
+                let path = candidateURL.absoluteString.replacingOccurrences(of: "file://", with: "")
+                let absolutePath = AbsolutePath(stringLiteral: path)
```
Comment (danyf90) [inline Sources/TuistLoader/Utils/ManifestFilesLocator.swift:112]: Doesn't URL have a `path` property?
```suggestion
                let absolutePath = AbsolutePath(stringLiteral: candidateURL.path)
```

**Review by danyf90 — APPROVED** [review]

File: Sources/TuistLoader/Utils/ManifestFilesLocator.swift [hunk 3]
```diff
@@ -106,12 +106,13 @@ public final class ManifestFilesLocator: ManifestFilesLocating {
 
         let enumerator = FileManager.default.enumerator(atPath: path.pathString)
         while let candidatePath = enumerator?.nextObject() as? String {
-            let candidateURL = URL(fileURLWithPath: candidatePath)
-            if fileNamesCandidates.contains(candidateURL.lastPathComponent) {
-                let path = candidateURL.absoluteString.replacingOccurrences(of: "file://", with: "")
-                let absolutePath = AbsolutePath(stringLiteral: path)
-                if hasValidManifestContent(absolutePath) {
-                    tuistManifestsFilePaths.append(absolutePath)
+            if candidatePath.hasSuffix(".swift") {
```
Comment (danyf90) [inline Sources/TuistLoader/Utils/ManifestFilesLocator.swift:109]: nit: you can use a guard and continue to reduce intentation

**Review by fortmarek — APPROVED** [review]
This is an amazing improvement 👏

**Review by pepicrft — COMMENTED** [review]
just wow. Thanks a lot for this @anlaital-oura

**Comment by pepicrft** [comment]
@all-contributors add @anlaital-oura for code

**Pull request merged**